[CLI] Add configuration parameters needed to specify multiple instance types for a compute resource

EddyMM opened the pull request:
### Description of changes
* As part of supporting multiple instance types in a compute resource, these configuration parameters are being added
```
Scheduling:
  Scheduler: slurm
  SlurmQueues:
  - Name: ...
    AllocationStrategy: lowest-price|capacity-optimized
    ComputeResources:
    - Name: ...
      InstanceTypeList:
        - InstanceType: ...
        - InstanceType: ...
```

- A Compute Resource may define a list of possible instance-types that can be launched based on an allocation strategy
- A queue may have compute resources that define a single instance type or a list of instance-types
- The AllocationStrategy is defined at Queue Level and will apply to all the compute fleets launched under that queue.

### Tests
* A queue can list all instance types it contains (including the compute resources with multiple instance types)
* A queue can support compute Resources containing a either Single Instance Type or Multiple Instance Types 
* For launch templates generated for a compute fleet with multiple instance types:
    * The smallest MaxNetworkCards value will be used to determine the number of network interfaces to use
    * The InstanceType and CpuInfo fields are excluded

### References
* [AllocationStrategy](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/ec2-fleet-allocation-strategy.html)

### Checklist
- [x] Make sure you are pointing to **the right branch** and add a label in the PR title (i.e. **2.x** vs **3.x**)
- [x] Check all commits' messages are clear, describing what and why vs how.
- [x] Make sure **to have added unit tests or integration tests** to cover the new/modified code.
- [ ] Check if documentation is impacted by this change.

Please review the [guidelines for contributing](../CONTRIBUTING.md) and [Pull Request Instructions](https://github.com/aws/aws-parallelcluster/wiki/Git-Pull-Request-Instructions).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

enrico-usai commented on cli/src/pcluster/config/cluster_config.py:
```diff
@@ -1614,12 +1614,11 @@ def scheduler_resources(self):
 
 
-class SlurmComputeResource(BaseComputeResource):
+class BaseSlurmComputeResource(BaseComputeResource):
```
minor: we can add a starting underscore to identify it's a class used within the module only.

enrico-usai commented on cli/src/pcluster/config/cluster_config.py:
```diff
@@ -1629,17 +1628,85 @@ def __init__(
+        self.instance_type = Resource.init_param(instance_type)
+
+
+def get_instance_type_info(instance_type) -> InstanceTypeInfo:
```
I'd avoid defining a function that is one line and with the same name of the called function, we can call the `get_instance_type_info` directly.

enrico-usai replied:
or we can define this function into the `SlurmFlexibleComputeResource` as static method given that it's used just by that class.

enrico-usai commented on cli/src/pcluster/config/cluster_config.py:
```diff
@@ -1629,17 +1628,85 @@ def __init__(
+        return [flexible_instance_type.instance_type for flexible_instance_type in self.instance_types]
+
+    @property
+    def disable_simultaneous_multithreading_manually(self) -> bool:
```
Here we're overriding methods from the parent class.

What about removing these methods from `BaseSlurmComputeResource` and have them defined just in the `SlurmFlexibleComputeResource` and  `SlurmComputeResource`.

Then we could mark `BaseSlurmComputeResource`  as abstract class and set the methods as `abstractmethods` to be sure they are defined by children classes.

enrico-usai commented on cli/src/pcluster/config/cluster_config.py:
```diff
@@ -1629,17 +1628,85 @@ def __init__(
+    @property
+    def is_ebs_optimized(self) -> bool:
+        """Return True if all the instances have optimized EBS support."""
+        for instance_type in self.instance_type_names:
```
nit: If I'm not wrong this can be rewritten with `any`

EddyMM replied:
I think [all](https://docs.python.org/3/library/functions.html#all) is what would work here.
I went for this approach to avoid trying to build an iterable/list in order to use `all`. 
But there might be a possibility to use a generator (or generator comprehension). I'll give it a try to see the results.

EddyMM replied:
Eventually decided to avoid using `all` here.
Using `all` works but it's not easily readable.

enrico-usai commented on cli/src/pcluster/schemas/cluster_schema.py:
```diff
@@ -1173,6 +1211,9 @@ class SlurmQueueSchema(_CommonQueueSchema):
     @post_load
     def make_resource(self, data, **kwargs):
         """Generate resource."""
+        allocation_strategy = data.get("allocation_strategy")
```
ComputeResources with InstanceTypes and InstanceType can coexist in the same Queue. It means we cannot have SlurmFlexibleQueue, but add all the logic to manage both of them to the SlurmQueue.

enrico-usai commented on cli/src/pcluster/schemas/cluster_schema.py:
```diff
@@ -1161,6 +1195,10 @@ class _CommonQueueSchema(BaseQueueSchema):
 
+    allocation_strategy = fields.Str(
+        validate=validate.OneOf(["lowest-price", "capacity-optimized"]),
+        metadata={"update_policy": UpdatePolicy.COMPUTE_FLEET_STOP},
```
Did we evaluate `QUEUE_UPDATE_STRATEGY`?

enrico-usai commented on cli/src/pcluster/config/cluster_config.py:
```diff
@@ -1796,6 +1860,39 @@ def _register_validators(self):
             )
 
+
+class SlurmFlexibleQueue(BaseSlurmQueue):
```
We cannot have a separate Queue class for flexible instance types. Both types of compute resources should coexist in the same Queue.

enrico-usai commented on cli/src/pcluster/config/cluster_config.py:
```diff
@@ -2373,8 +2485,13 @@ def get_instance_types_data(self):
+                    instance_type_info = compute_resource.instance_type_info
+                    result[instance_type_info.instance_type()] = instance_type_info.instance_type_data
+                if isinstance(compute_resource, SlurmFlexibleComputeResource):
+                    for instance_type in compute_resource.instance_type_names:
```
can we generalize the compute_resource interface to return a list of instance_type_names for both the cases?

enrico-usai replied:
I'd add a property called:  `instance_type_list` in the base compute resource (like the one we have in the Queue), that returns a list with an item in the basic case or a list of instance types in the flexible case.

enrico-usai replied:
I'm for calling the function "instance_type_list" and not "instance_type_names", to avoid introducing another naming. We're already using "instance_type_list" at queue level so we can have the same at compute resource level.

lukeseawalker commented on cli/src/pcluster/schemas/cluster_schema.py:
```diff
@@ -1082,7 +1095,12 @@ class _ComputeResourceSchema(BaseSchema):
+    instance_types = fields.Nested(
+        FlexibleInstanceTypeSchema,
+        many=True,
+        metadata={"update_policy": UpdatePolicy.COMPUTE_FLEET_STOP, "update_key": "InstanceType"},
```
UpdatePolicy.COMPUTE_FLEET_STOP_ON_REMOVE

lukeseawalker commented on cli/src/pcluster/schemas/cluster_schema.py:
```diff
@@ -1082,7 +1095,12 @@ class _ComputeResourceSchema(BaseSchema):
-    instance_type = fields.Str(required=True, metadata={"update_policy": UpdatePolicy.COMPUTE_FLEET_STOP})
+    instance_type = fields.Str(metadata={"update_policy": UpdatePolicy.COMPUTE_FLEET_STOP})
+    instance_types = fields.Nested(
+        FlexibleInstanceTypeSchema,
```
What about going for "InstanceTypeSchema"? I think it's simpler

EddyMM replied:
I initially used `InstanceTypeSchema` but figured it might be confused with the properties in an InstanceTypeInfo object.
But since this schema is used in the context of a FlexibleComputeResource, I can simply use `InstanceTypeSchema`.

lukeseawalker commented on cli/src/pcluster/config/cluster_config.py:
```diff
@@ -1629,17 +1629,127 @@ def __init__(
+        In this case the Compute Resource may have multiple instance types, hence the instance-type with
+        the least MaxNetworkInterfaceCards value will be considered.
+        """
+        least_max_nics = SlurmFlexibleComputeResource.get_instance_type_info(
```
if this method works for a instance type list with just 1 element, I will define it in the Base class so that it can be used for both SlurmFlexibleComputeResource and SlurmComputeResource

lukeseawalker commented on cli/src/pcluster/config/cluster_config.py:
```diff
@@ -1629,17 +1629,127 @@ def __init__(
+        return least_max_nics
+
+    @property
+    def is_ebs_optimized(self) -> bool:
```
same here, this can be defined in the Base class

lukeseawalker commented on cli/src/pcluster/config/cluster_config.py:
```diff
@@ -2284,16 +2403,17 @@ def _register_validators(self):
+                    )
+                    self._register_validator(
+                        InstanceArchitectureCompatibilityValidator,
+                        instance_type=instance_type,
```
let's think about passing directly a list of InstanceTypeInfo and we move the architecture method in there

lukeseawalker commented on cli/src/pcluster/config/cluster_config.py:
```diff
@@ -1755,17 +1865,37 @@ def queue_ami(self):
+        """Return a set of instance types in the queue that have instance store."""
+        result = set()
+        for instance_type in self.instance_type_list:
+            if SlurmFlexibleComputeResource.get_instance_type_info(instance_type).instance_storage_supported():
```
I think we shouldn't use the SlurmFlexibleComputeResource here.
What about having a method to get the list of instance type info, something like instance_type_info_list?
This method will be exposed by the compute_resource, by both the SlurmFlexibleComputeResource and SlurmComputeResource, but in this way, the caller (instance_types_with_instance_storage) doesn't have to know about them.

EddyMM replied:
I agree, the `Queue` Class shouldn't rely on the `get_instance_type_info` method from `SlurmFlexibleComputeResource`.
I'll push the responsibility to obtain the list of instance_types that have instance store down to the compute resources as suggested.

lukeseawalker commented on cli/src/pcluster/config/cluster_config.py:
```diff
@@ -2373,8 +2493,9 @@ def get_instance_types_data(self):
-                instance_type_info = compute_resource.instance_type_info
-                result[instance_type_info.instance_type()] = instance_type_info.instance_type_data
+                for instance_type in compute_resource.instance_types:
+                    instance_type_info = SlurmFlexibleComputeResource.get_instance_type_info(instance_type)
```
same consideration here, compute_resource should have a method to retrieve the instance_type_info_list

lukeseawalker commented on cli/tests/pcluster/config/test_cluster_config.py:
```diff
@@ -60,6 +62,29 @@ def base_slurm_cluster_config(self):
+            compute_resources=[
+                SlurmComputeResource(name="compute_resource_1", instance_type="c5n.4xlarge"),
+                SlurmComputeResource(name="compute_resource_2", instance_type="t2.micro"),
+                SlurmFlexibleComputeResource(
```
add a SlurmFlexibleComputeResource with just one element

enrico-usai commented on CHANGELOG.md:
```diff
@@ -8,6 +8,7 @@ CHANGELOG
 - Remove support for Python 3.6 in aws-parallelcluster-batch-cli.
 - Upgrade Python and NodeJS versions in API infrastructure, API Docker container and cluster Lambda resources.
 - Move head node tags from launch template to instance definition to avoid head node replacement on tags updates.
+- Support using multiple instance types in a compute resource based on an allocation strategy
```
I'd write: Add support for multiple instance types in the same Compute Resource.
Anyway missing final period.

enrico-usai commented on cli/src/pcluster/aws/aws_resources.py:
```diff
@@ -236,6 +236,19 @@ def is_ebs_optimized(self):
+        ebs_info = self.instance_type_data.get("EbsInfo")
+        if ebs_info:
+            return ebs_info.get("EbsOptimizedSupport")
+        return "unsupported"
```
an alternative one-liner way is:
```python
return self.instance_type_data.get("EbsInfo", {}).get("EbsOptimizedSupport", "unsupported")
```

enrico-usai replied:
Anyway do we really need this or can we use the existing `is_ebs_optimized` ?

If we need both let's call `ebs_optimized_support` from `is_ebs_opmitized`

enrico-usai commented on cli/src/pcluster/config/cluster_config.py:
```diff
@@ -1629,16 +1629,140 @@ def __init__(
+
+    @property
+    def is_ebs_optimized(self) -> bool:
+        if "unsupported" in [
```
Why did you define `ebs_optimized_support()`? We already had `is_ebs_optimized` in the instance_type_info class that is doing the check for `unsupported` for you.

enrico-usai replied:
Then you can have an one-liner check:
```python
return all([self.instance_type_info_map[instance_type].is_ebs_optimized() for instance_type in self.instance_types])

EddyMM replied:
I agree, we can reuse the `is_ebs_optimized` method from InstanceTypeInfo to find out EBS-Support at ComputeResource level.
I used `ebs_optimized_support()` originally to separate the logic used to determine EBS-Support followed at ComputeResource level vs InstanceType level.

enrico-usai commented on cli/src/pcluster/config/cluster_config.py:
```diff
@@ -1629,16 +1629,140 @@ def __init__(
+        the least MaxNetworkInterfaceCards value will be considered.
+        """
+        least_max_nics = self.instance_type_info_map[self.instance_types[0]].max_network_interface_count()
+        if len(self.instance_types) > 1:
```
let's add a unit test for this property calculation

EddyMM replied:
This is tested in the test_cluster_stack: [#L235-L235](https://github.com/EddyMM/aws-parallelcluster/blob/wip/multiple-instance-types/cli/tests/pcluster/templates/test_cluster_stack.py#L235-L235)
It selects the least number of NICs in the list.

enrico-usai commented on cli/src/pcluster/config/cluster_config.py:
```diff
@@ -1755,17 +1874,36 @@ def queue_ami(self):
+    def instance_types_with_instance_storage(self):
+        """Return a set of instance types in the queue that have instance store."""
+        result = set()
+        for compute_resource in self.compute_resources:
```
you can use set comprehension:
```python
return {compute_resource.instance_types_with_instance_storage for compute_resource in self.compute_resources}
```

enrico-usai replied:
same in the instance_type_list property.

EddyMM replied:
`compute_resource.instance_types_with_instance_storage` and `compute_resource.instance_types` return a **list** of instance_types so set comprehension won't work since lists are unhashable.

enrico-usai commented on cli/src/pcluster/schemas/cluster_schema.py:
```diff
@@ -1092,9 +1110,24 @@ class SlurmComputeResourceSchema(_ComputeResourceSchema):
+            **kwargs,
+        ):
+            raise ValidationError(
+                "A Compute Resource needs to specify either a single InstanceType or a list of InstanceTypes."
```
a ComputeResource ... or InstanceTypesList

lukeseawalker replied:
docstring not updated

enrico-usai commented on cli/src/pcluster/config/cluster_config.py:
```diff
@@ -1629,16 +1629,140 @@ def __init__(
         self.efa = efa or Efa(enabled=False, implied=True)
         self.schedulable_memory = Resource.init_param(schedulable_memory)
+        self.__instance_types_with_instance_storage = []
+        self.__instance_type_info_map = {}
```
why are you using two underscore? I'd stay consistent with the rest of the code using a single underscore for the private attributes.

enrico-usai commented on cli/src/pcluster/config/cluster_config.py:
```diff
@@ -1629,16 +1629,140 @@ def __init__(
+    @property
+    def instance_types_with_instance_storage(self):
+        if not self.__instance_types_with_instance_storage:
+            for instance_type in self.instance_types:
```
nit: you can use list comprehension

enrico-usai commented on cli/tests/pcluster/templates/test_cluster_stack.py:
```diff
@@ -176,6 +177,100 @@ def test_scheduler_plugin_substack(mocker, config_file_name, expected_scheduler_
+    )
+
+    instance_type = get_launch_template_data_property("InstanceType")
+    assert bool(instance_type) == has_instance_type
```
you can use `assert_that` from pytest for all the check you're doing.

enrico-usai commented on cli/tests/pcluster/validators/test_all_validators.py:
```diff
@@ -65,6 +65,20 @@ def _load_and_validate(config_path):
+        architecture = kwargs.get("architecture")
+        expected_instance_type_list = expected_instance_architecture_validator_input[call_index].get("instance_types")
+        expected_architecture = expected_instance_architecture_validator_input[call_index].get("architecture")
+        assert len(instance_type_list) == len(expected_instance_type_list)
```
you can use:
```python
assert_that(instance_type_list).is_length(len(expected_instance_type_list))
```

enrico-usai commented on cli/tests/pcluster/validators/test_all_validators.py:
```diff
@@ -65,6 +65,20 @@ def _load_and_validate(config_path):
+        expected_architecture = expected_instance_architecture_validator_input[call_index].get("architecture")
+        assert len(instance_type_list) == len(expected_instance_type_list)
+        assert len(set(instance_type_list) - set(expected_instance_type_list)) == 0
+        assert architecture == expected_architecture
```
assert_that(...).is_equal_to(...)

lukeseawalker commented on cli/src/pcluster/aws/ec2.py:
```diff
@@ -244,9 +244,13 @@ def get_instance_type_info(self, instance_type):
+        """Return a list that filters out unsupported architectures that an instance type may have."""
+        supported_architectures = instance_type_info.supported_architecture()
+        # Some instance types support multiple architectures (x86_64 and i386). Filter out unsupported ones.
         return list(set(supported_architectures) & set(SUPPORTED_ARCHITECTURES))
```
isn't this filtering out already done in InstanceTypeInfo.supported_architecture()?

lukeseawalker commented on cli/src/pcluster/config/cluster_config.py:
```diff
@@ -1629,16 +1629,140 @@ def __init__(
+        return self.disable_simultaneous_multithreading
+
+    @property
+    def disable_simultaneous_multithreading_via_cpu_options(self) -> bool:
```
this and following property have been removed as part of https://github.com/aws/aws-parallelcluster/pull/4260

lukeseawalker replied:
you can remove the `disable_simultaneous_multithreading_via_cpu_options` method

lukeseawalker commented on cli/src/pcluster/schemas/cluster_schema.py:
```diff
@@ -1051,6 +1053,17 @@ def make_resource(self, data, **kwargs):
+    """Schema of a compute resource that supports a pool of instance types."""
+
+    instance_type = fields.Str(required=True, metadata={"update_policy": UpdatePolicy.COMPUTE_FLEET_STOP})
+
```
merge conflict

lukeseawalker commented on cli/src/pcluster/schemas/cluster_schema.py:
```diff
@@ -1092,9 +1110,24 @@ class SlurmComputeResourceSchema(_ComputeResourceSchema):
         """Generate resource."""
+        if data.get("instance_type_list"):
+            return SlurmFlexibleComputeResource(**data)
         return SlurmComputeResource(**data)
```
What about "A Compute Resource needs to specify either InstanceType or InstanceTypeList."

lukeseawalker commented on cli/src/pcluster/templates/cluster_stack.py:
```diff
@@ -1392,25 +1394,33 @@ def _add_compute_resource_launch_template(
+
+        if compute_resource.is_ebs_optimized:
+            conditional_template_properties.update({"ebs_optimized": True})
+        if compute_resource.pass_cpu_options_in_launch_template:
```
cpu_options dropped in https://github.com/aws/aws-parallelcluster/pull/4260

lukeseawalker commented on cli/src/pcluster/aws/aws_resources.py:
```diff
@@ -221,11 +221,17 @@ def instance_type(self):
-        if ebs_info:
-            ebs_optimized = ebs_info.get("EbsOptimizedSupport") != "unsupported"
-        return ebs_optimized
+        return self.ebs_optimized_support() != "unsupported"
```
cannot we leave just `def is_ebs_optimized`?

lukeseawalker commented on cli/src/pcluster/schemas/cluster_schema.py:
```diff
@@ -1092,9 +1110,22 @@ class SlurmComputeResourceSchema(_ComputeResourceSchema):
 
+    @validates_schema
+    def no_coexist_instance_type_flexibility(self, data, **kwargs):
+        """Validate that 'instance-type' and 'instance-types' do not co-exist."""
```
fix doc string, e.g.
```
"""Validate that 'instance_type' and 'instance_type_list' do not co-exist."""
```

lukeseawalker commented on cli/src/pcluster/validators/cluster_validators.py:
```diff
@@ -234,17 +237,22 @@ class InstanceArchitectureCompatibilityValidator(Validator):
-                FailureLevel.ERROR,
-            )
+
+        compute_architectures = set()
```
the `set` is not needed because you need to compare head_node to each_compute resource and not head_node to the incremental set of compute_resources

lukeseawalker commented on cli/tests/pcluster/schemas/test_cluster_schema.py:
```diff
@@ -316,6 +318,104 @@ def test_scheduling_schema(mocker, config_dict, failure_message):
+            },
+            "",
+        ),
+        # Failing to specify either InstanceType or InstanceTypes should return a validation error
```
add test case where both (InstanceType or InstanceTypeList) are specified (for the same compute resource), this should return a validation error

EDIT, I saw it is covered in test_slurm_flexible_compute_resources. Is it possible to merge the two tests?

lukeseawalker replied:
also fix comment InstanceTypes -> InstanceType or InstanceTypeList

lukeseawalker commented on cli/tests/pcluster/templates/test_cluster_stack.py:
```diff
@@ -176,6 +177,100 @@ def test_scheduler_plugin_substack(mocker, config_file_name, expected_scheduler_
+    config_file_name,
+    has_instance_type,
+    no_of_network_interfaces,
+    includes_cpu_options,
```
includes_cpu_options not used

lukeseawalker commented on cli/src/pcluster/config/cluster_config.py:
```diff
@@ -1600,16 +1600,131 @@ def __init__(
+        self._instance_types_with_instance_storage = []
+        self._instance_type_info_map = {}
+
+    @staticmethod
```
can this be a property?

lukeseawalker commented on cli/src/pcluster/config/cluster_config.py:
```diff
@@ -1600,16 +1600,131 @@ def __init__(
+
+
+class FlexibleInstanceType(Resource):
+    """Represent an instance type listed under InstanceTypes under ComputeResources."""
```
InstanceTypes -> InstanceTypeList

lukeseawalker commented on cli/src/pcluster/config/cluster_config.py:
```diff
@@ -1600,16 +1600,131 @@ def __init__(
+        return False
+
+    @property
+    def pass_cpu_options_in_launch_template(self) -> bool:
```
you can remove this method, isn't used anywhere

lukeseawalker commented on cli/src/pcluster/config/cluster_config.py:
```diff
@@ -2226,14 +2349,15 @@ def _register_validators(self):
                 self._register_validator(
                     InstanceArchitectureCompatibilityValidator,
-                    instance_type=compute_resource.instance_type,
+                    instance_type_info_list=list(compute_resource.instance_type_info_map.values()),
```
is `InstanceArchitectureCompatibilityValidator` accepting a instance_type_info_list ?

lukeseawalker replied:
Let's keep passing `instance_type` for now (this require to add a for loop to iterate of the compute resource instance type

lukeseawalker replied:
nevermind, I saw that InstanceArchitectureCompatibilityValidator was modified to reflect the changes

EddyMM replied:
Yes.

EddyMM committed:
[CLI] Add configuration parameters needed to specify multiple instance types for a compute resource

- A Compute Resource may define a list of possible instance-types that can be launched based on an allocation strategy
- A queue can have compute resources that define a single instance type or a list of instance-types
- The AllocationStrategy is defined at Queue Level and will apply to all the compute fleets launched under that queue.

Signed-off-by: Eddy Mwiti <eddmwiti@amazon.com>

Q-Hilbtec commented:
This is great, thanks @lukeseawalker! I noted that it is not currently compatible with Memory Based Scheduling - would it be possible to loosen the constraint and allow users to specify multiple instance types - assuming all of them share the same RAM amount?
